GoMeet/src/main/webapp/js/meeting-time-dao.js

```diff
@@ -0,0 +1,86 @@
+class MeetingTimeDAO {
+  /** Data Access Object for MeetingTimes */
+
+  static endpoint = '/meeting-time?'; // URL endpoint of the associated servlet
+  
+  /**
+   * Returns a URL string with the key value pairs
+   * as given in 'keyValues' encoded into a query string, in the format
+   * '/meeting-time?key1=value1&key2=value2...'
+   * @param {Object} fieldValues the map of field to values representing the
+   * field-value pairs to be appended to the querystring.
+   * @returns the url at the /meeting-time endpoint with the query string
+   * appended
+   */
+  static url(fieldValues) {
+    let queryString = new URLSearchParams();
+    for (const [field, value] of Object.entries(fieldValues)) {
+      queryString.append(field, value);
+    }
+    return MeetingTimeDAO.endpoint + queryString.toString();
+  }
+  
+  /**
+   * Fetches the meeting time data associated with the provided
+   * MeetingTimeID.
+   * @param {String} meetingTimeId the MeetingTimeID of the MeetingTime 
+   * entity to be fetched from Datastore. Equivalent to the 
+   * Datastore-generated Key of the entity.
+   * @returns a JSON object with the details of the fetched 
+   * MeetingTime entity.
+   */
+  static async fetchMeetingTime(meetingTimeId) {
+    if (meetingTimeId === null || meetingTimeId === undefined) {
+      throw new Error(INSUFFICIENT_REQUEST_PARAM);
+    }
+     
+    if (typeof meetingTimeId !== 'string') {
+      throw new Error(INVALID_PARAM_TYPE);
+    }
+
+    let urlString = MeetingTimeDAO.url({'meetingTimeId': meetingTimeId});
+    let results = await fetch(urlString).then((results) => results.json());
+    return results;
+  }
+
+  /**
+   * Completes a POST request to create a new MeetingTime entity
+   * with the data as provided in datetimeStr and saves the entity to
+   * Datastore.
+   * @param {String} datetimeStr the datetime string in the exact format
+   * YYYY-MM-DDTHH:MM, representing the date and time to be stored in the 
+   * new MeetingTime Entity
+   * @return JSON containing the meetingTimeId of the newly created 
+   * MeetingTime entity. 
+   */
+  static async newMeetingTime(datetimeStr) {
+    if (datetimeStr === null || datetimeStr === undefined) {
+      throw new Error(INSUFFICIENT_REQUEST_PARAM);
+    }
+
+    if (typeof datetimeStr !== 'string') {
+      throw new Error(INVALID_PARAM_TYPE);
+    }
+
+    // check that the datetime string is in the format: YYY-MM-DDTHH:MM
+    let format = new RegExp(/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d/);
+    if (!format.test(datetimeStr)) {
+      throw new Error(INVALID_PARAM_VALUE);
+    } 
```

Comment:
Line 73-76 already check the format of the string, any reason to keep this? It also seems that the regex will accept "2020-15-39T28:00" as a match, so maybe consider removing this check?

Reply:
The Date constructor can take a number of different formats (e.g. "20 Jan 2021" can also initialise a date), but the _only_ type of format I want here is the '2021-01-20T11:40'. 

Yep, it is because the regex will accept invalid dates like the one you mentioned, after I check the format, I need pass the string _again_ to the Date constructor to check that it is a valid date _as well as_ being of the correct format.

Hopefully that makes sense!

Reply:
The regex here is quite scary, and doesn't actually do its job well (couldn't check the content of the string by itself and still need to depends on Date() constructor, and in fact, it'll match something like `blah blah blah 2020-15-39T28:00 blah blah blah`). 

Can you use something like `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$` which will strictly use to match the format of the string? Also could you update the comment as well (you missed a Y)?

Reply:
Yep, I agree - the additional checks are really not necessary if I'm going to pass it to the Date constructor anyway. I have changed it now 👍